File: problems/树/572.另一棵树的子树.py
#!/usr/bin/env python3
"""
572. 另一棵树的子树
"""
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def is_subtree(root: Optional[TreeNode], sub_root: Optional[TreeNode]) -> bool:
    if not root: return False
    def is_same(t1: Optional[TreeNode], t2: Optional[TreeNode]) -> bool:
        if not t1 and not t2: return True
        if not t1 or not t2: return False
        return t1.val == t2.val and is_same(t1.left, t2.left) and is_same(t1.right, t2.right)
    if is_same(root, sub_root): return True
    return is_subtree(root.left, sub_root) or is_subtree(root.right, sub_root)

if __name__ == "__main__":
    root = TreeNode(3, TreeNode(4, TreeNode(1), TreeNode(2)), TreeNode(5))
    sub = TreeNode(4, TreeNode(1), TreeNode(2))
    print(is_subtree(root, sub))  # True
    sub2 = TreeNode(3, TreeNode(1), TreeNode(2))
    print(is_subtree(root, sub2))  # False
    print(is_subtree(None, None))  # False

File: problems/树/572.另一棵树的子树.cpp
#include <iostream>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

bool isSameTree(TreeNode* a, TreeNode* b) {
    if (!a && !b) return true;
    if (!a || !b) return false;
    return a->val == b->val && isSameTree(a->left, b->left) && isSameTree(a->right, b->right);
}
bool isSubtree(TreeNode* root, TreeNode* subRoot) {
    if (!root) return false;
    if (isSameTree(root, subRoot)) return true;
    return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
}
int main() {
    TreeNode* root = new TreeNode(3);
    root->left  = new TreeNode(4);
    root->right = new TreeNode(5);
    root->left->left  = new TreeNode(1);
    root->left->right = new TreeNode(2);
    TreeNode* sub = new TreeNode(4);
    sub->left  = new TreeNode(1);
    sub->right = new TreeNode(2);
    cout << (isSubtree(root, sub) ? "True" : "False") << endl;  // True
    TreeNode* sub2 = new TreeNode(3);
    sub2->left  = new TreeNode(1);
    sub2->right = new TreeNode(2);
    cout << (isSubtree(root, sub2) ? "True" : "False") << endl;  // False
    cout << (isSubtree(nullptr, nullptr) ? "True" : "False") << endl;  // False
    return 0;
}
